XRPLib/imu.py

```diff
@@ -321,6 +340,30 @@ def gyro_scale(self, dat=None):
             else: return
             self._r_w_reg(LSM6DSO_CTRL2_G, self._scale_g<<1, 0xF1)
 
+    def acc_rate(self, dat=None):
+        """
+        Set the accelerometer rate. The rate can be '0', '12.5', '26', '52', '104', '208', '416', '833', '1660', '3330', '6660'.
```

Comment:
First, to answer your questions - I was really just copying the structure of the scale functions. I'm actually not really keen on doing it like this; it's slow, takes more memory than it needs to, and it's not really how it should be done. Using numbers would be an improvement, but I think there's better solutions (eg. an enum).

Ok, this is where I tangent. This might turn into a longer discussion about the best way to architect the library. If so, it might be worth moving this conversation to its own issue in the repo. Sorry in advance 😉

Quick disclaimer: I mostly write C/C++ code (usually Arduino). There's actually a lot of things about Python that drive me crazy, so I don't use Python much. That means I'm also ignorant about how Python works under the hood, and I don't know a lot of features that I probably should. With that said:

If I were doing this in C/C++, there's some things I'd do differently:
1. Use enums for function parameters
    * Create one enum with each valid option for each function, then use those enums as the function parameters. That guarantees a correct value is passed at compile time, rather than having to check at runtime (faster execution, less program memory)
2. Use `#define`s for register names/addresses
    * This puts all those values into program memory rather than storing them in global variables, so this saves on RAM. Also, the compiler would just ignore any unused values, which further saves program memory
3. Use bitfield structs for register values ([explanation](https://en.cppreference.com/w/cpp/language/bit_field) for anyone who needs it)
    * This basically does all the bit shifting and masking for you. It guarantees that you don't accidentally set any bits that you didn't mean to, and actually results in faster code on some architectures

Again, I'm ignorant about Python, so I don't know the best practices for this kind of stuff. But after some Googling and testing, I have some thoughts:
1. Looks like enums currently [aren't support in MicroPython](https://github.com/micropython/micropython-lib/issues/269)
    * I'm not sure what the best alternative would be. I think the tuple solution that's being used now is interesting, where the index is the value for each entry. But I don't think it's very scalable, it's difficult to verify the indices are correct, and it requires adding blank entries to skip values. I think something like a dictionary might make more sense if we can't use enums.
    * This unfortunately doesn't guarantee the provided value is valid (insert grumbling about dynamic type casting and no compile time checks in Python...). Regardless of what we expect the parameters to be, it would be wise to validate all parameters of all functions (at least, any functions that users would call)
2. Use [MicroPython's `const`](https://docs.micropython.org/en/latest/library/micropython.html#micropython.const) for register names/addresses
    * This should help with RAM usage. Don't think it'll help much with program memory (arguably worse because there's more text in the .py file, but that's another conversation...), but probably worth it
3. Use [MicroPython's `uctype` bit field structs](https://docs.micropython.org/en/latest/library/uctypes.html) for register values
    * I've been tinkering with this a bit, and it behaves like I expect. It's not documented as well as I'd like, but it's relatively straight forward I think.

If you need, I might be able to provide an example that includes the ideas I'm suggesting here. Let me know your thoughts!

Reply:
i agree with a lot of what you said here; I've used pretty much this exact IMU in C++, and i did basically that (except the bitfields, which i've never messed with). 

In regards to the Micropython stuff, enums not being supported is pretty annoying. There's definitely still ways around that though, and I'm tempted to go and try to make that work anyways. I definitely agree with parameter validation for functions where the type of which isn't explicitly defined (or at least add much better documentation). 

I had originally had it set up using `const`, but that was causing issues with Sphinx and the auto-documentation generation. If RAM usage winds up being an issue I can go back and revisit trying to get that to work.

As I mentioned, i haven't used bitfields before so that'll be new, but it definitely seems like a worthwhile thing to look into in terms of cleaning up this code a little. 

Overall I just want to make sure that the API is as easy to use as possible. I really would like to at the very least shift over to an enum-like situation, even if it's not as efficient as it would be had micropython allowed it. I'm thinking maybe just a dictionary would be a little better?

Let me know if you want to give it a shot, otherwise I'm fine merging your stuff in now and having me go back to clean up everything later

Reply:
Ok, so here's an example of how I think we could improve the structure of the IMU code, implementing the 3 things I mentioned (`const`, dictionaries, and bit field structs). This should be pretty easy to scale up as needed. This code here is just a bunch of definitions, so it'd maybe make sense to put it in its own `imu_defs.py` file or something:

```
from uctypes import BFUINT8, BF_POS, BF_LEN, struct, addressof
from micropython import const

"""
	Register addresses
"""
LSM_REG_CTRL1_XL = const(0x10)
LSM_REG_CTRL2_G  = const(0x11)
LSM_REG_CTRL3_C  = const(0x12)
LSM_REG_CTRL4_C  = const(0x13)
LSM_REG_CTRL5_C  = const(0x14)
LSM_REG_CTRL6_C  = const(0x15)
LSM_REG_CTRL7_G  = const(0x16)
LSM_REG_CTRL8_XL = const(0x17)
LSM_REG_CTRL9_XL = const(0x18)
LSM_REG_CTRL10_C = const(0x19)

"""
	Dictionaries for possible register settings
"""
LSM_ODR = {
	"0Hz"    : 0x0,
	"12.5Hz" : 0x1,
	"26Hz"   : 0x2,
	"52Hz"   : 0x3,
	"104Hz"  : 0x4,
	"208Hz"  : 0x5,
	"416Hz"  : 0x6,
	"833Hz"  : 0x7,
	"1660Hz" : 0x8,
	"3330Hz" : 0x9,
	"6660Hz" : 0xA,
}
LSM_ACCEL_FS = {
	"2g"  : 0x0,
	"16g" : 0x1,
	"4g"  : 0x2,
	"8g"  : 0x3,
}
LSM_GYRO_FS = {
	"125dps"  : 0x1,
	"250dps"  : 0x0,
	"500dps"  : 0x2,
	"1000dps" : 0x4,
	"2000dps" : 0x6,
}

"""
	Bit field struct definitions of registers
"""
LSM_REG_CTRL1_XL_LAYOUT = {
    "LPF2_XL_EN" : BFUINT8 | 1 << BF_POS | 1 << BF_LEN,
    "FS_XL"      : BFUINT8 | 2 << BF_POS | 2 << BF_LEN,
    "ODR_XL"     : BFUINT8 | 4 << BF_POS | 4 << BF_LEN,
}
LSM_REG_CTRL2_G_LAYOUT = {
    "FS_G"  : BFUINT8 | 1 << BF_POS | 3 << BF_LEN,
    "ODR_G" : BFUINT8 | 4 << BF_POS | 4 << BF_LEN,
}

"""
	Copies of register values
"""
reg_ctrl1_xl_byte   = bytearray(1)
reg_ctrl1_xl_struct = struct(addressof(reg_ctrl1_xl_byte), LSM_REG_CTRL1_XL_LAYOUT)
reg_ctrl2_g_byte   = bytearray(1)
reg_ctrl2_g_struct = struct(addressof(reg_ctrl2_g_byte), LSM_REG_CTRL2_G_LAYOUT)
```

Obviously this doesn't include everything, just getting the idea across. If I were doing this in C/C++, I'd add everything for every register. However we need to keep in mind file size, since all of this text gets stored on the Pico W, which only has 2MB of flash. So it would probably be better in this case to only add the things we actually use.

Then in the IMU class, the member functions would look more like this:

```
def acc_rate(self, dat=None):
	"""
	Set the accelerometer rate. The rate can be '0Hz', '12.5Hz', '26Hz', '52Hz', '104Hz', '208Hz', '416Hz', '833Hz', '1660Hz', '3330Hz', '6660Hz'.
	Pass in no parameters to retrieve the current value
	"""
	# Check if dat is valid input
	if (dat is None) or (type(dat) is not str) or (dat not in LSM_ODR):
		# Not valid input, return ODR value
		reg_ctrl1_xl_byte = self._getreg(LSM_REG_CTRL1_XL)
		return reg_ctrl1_xl_struct.ODR_XL
	else:
		# Valid input, update ODR value
		reg_ctrl1_xl_byte = self._getreg(LSM_REG_CTRL1_XL)
		reg_ctrl1_xl_struct.ODR_XL = LSM_ODR[dat]
		return self._setreg(LSM_REG_CTRL1_XL, reg_ctrl1_xl_byte)
```

Note - I've not actually tested any of this, don't have my robot with me right now.

There's definitely pros and cons to this approach. The benefit is that this is pretty straight forward to scale up, there's no magic numbers, and it's quite readable IMO. The downside is that this takes up more memory, both flash and RAM (don't think dictionaries work with `const`). There's probably other considerations I'm not thinking of right now.

Let me know what you think!

Reply:
How much of the flash memory does the current library actually take up? I can't imagine adding the dictionaries would have a substantial impact there. I think in terms of clean implementation, I like the dictionaries a lot better than the current solution, since at least then the code becomes readable for any student looking to go above and beyond and try to understand how this stuff works (expecting this moreso from the WPI students using these in a class, not for the high schoolers). I also agree with moving these into a new file to keep `imu.py` clean.

If you want to implement that stuff as part of this PR, that would be awesome

Reply:
I believe the library currently uses about 200kB, including phew. I just downloaded the latest version of MicroPython for the Pico W, and it's 1368kB, so that presumably means we're using about 75% of the 2MB flash. We'll need to make sure there's extra room for user code, plus any updates to MicroPython and this library. Adding every definition for every register can definitely take a lot of space; [this file from ST's API](https://github.com/stm32duino/LSM6DSO/blob/main/src/lsm6dso_reg.h) is over 100kB of text, for example. Obviously we don't need to go all out like that, but I just want to make sure we're not being egregious with the amount of storage we take up.

I agree, I definitely prefer the code to be clean at the cost of some extra storage space. I'll start implementing this new code structure, and we'll see how much space it actually uses up. If needed, I think we can just delete any portions that we're not using to save space.

Reply:
Just pushed this new structure. Haven't tested it yet, so hopefully it's not completely broken 😉 I decided to only include the registers we're actually using, since adding everything increases both file size and RAM usage (more grumbling about Python...). But it should be easy to expand as needed.

Let me know what you think!

Reply:
I definitely like the organization of the imu_defs file a lot more than just having constants at the top of the imu class, thanks!

and yea, I don't anticipate really needing access to every single register option, what's there now should suffice.